In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/best-selling-books/best-selling-books.csv
/kaggle/input/30000-spotify-songs/readme.md
/kaggle/input/30000-spotify-songs/spotify_songs.csv
/kaggle/input/test-file/tested.csv
/kaggle/input/adult-census-income-dataset/adult.csv


# Label enconding

## Ingresos de personas Adultas

In [3]:
adult_census_df = pd.read_csv('/kaggle/input/adult-census-income-dataset/adult.csv')
adult_census_df.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

Predecir que personas podrian cobrar más de 50K

In [4]:
adult_census_df.describe()

age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [5]:
adult_census_df = adult_census_df.drop(['age'], axis=1) # extended index
adult_census_df = adult_census_df.drop(['fnlwgt'], axis=1) # extended index, not relevant
adult_census_df = adult_census_df.drop(['capital.gain'], axis=1) # extended index
adult_census_df = adult_census_df.drop(['capital.loss'], axis=1) # extended index
adult_census_df = adult_census_df.drop(['hours.per.week'], axis=1) # extended index
adult_census_df = adult_census_df.drop(['education'], axis=1) # education.num already fulfills the function
adult_census_df

workclass  education.num      marital.status         occupation  \
0             ?              9             Widowed                  ?   
1       Private              9             Widowed    Exec-managerial   
2             ?             10             Widowed                  ?   
3       Private              4            Divorced  Machine-op-inspct   
4       Private             10           Separated     Prof-specialty   
...         ...            ...                 ...                ...   
32556   Private             10       Never-married    Protective-serv   
32557   Private             12  Married-civ-spouse       Tech-support   
32558   Private              9  Married-civ-spouse  Machine-op-inspct   
32559   Private              9             Widowed       Adm-clerical   
32560   Private              9       Never-married       Adm-clerical   

        relationship   race     sex native.country income  
0      Not-in-family  White  Female  United-States  <=50K  
1      Not-in-family  White  Female  United-States  <=50K  
2          Unmarried  Black  Female  United-States  <=50K  
3          Unmarried  White  Female  United-States  <=50K  
4          Own-child  White  Female  United-States  <=50K  
...              ...    ...     ...            ...    ...  
32556  Not-in-family  White    Male  United-States  <=50K  
32557           Wife  White  Female  United-States  <=50K  
32558        Husband  White    Male  United-States   >50K  
32559      Unmarried  White  Female  United-States  <=50K  
32560      Own-child  White    Male  United-States  <=50K  

[32561 rows x 9 columns]

In [6]:
x=adult_census_df

In [7]:
%%time

adult_encoding=pd.DataFrame()
label=LabelEncoder()
for c in  x.columns:
    if(x[c].dtype=='object'):
        adult_encoding[c]=label.fit_transform(x[c])
    else:
        adult_encoding[c]=x[c]
        
adult_encoding

CPU times: user 92.2 ms, sys: 2.6 ms, total: 94.8 ms
Wall time: 94.1 ms


workclass  education.num  marital.status  occupation  relationship  \
0              0              9               6           0             1   
1              4              9               6           4             1   
2              0             10               6           0             4   
3              4              4               0           7             4   
4              4             10               5          10             3   
...          ...            ...             ...         ...           ...   
32556          4             10               4          11             1   
32557          4             12               2          13             5   
32558          4              9               2           7             0   
32559          4              9               6           1             4   
32560          4              9               4           1             3   

       race  sex  native.country  income  
0         4    0              39       0  
1         4    0              39       0  
2         2    0              39       0  
3         4    0              39       0  
4         4    0              39       0  
...     ...  ...             ...     ...  
32556     4    1              39       0  
32557     4    0              39       0  
32558     4    1              39       1  
32559     4    0              39       0  
32560     4    1              39       0  

[32561 rows x 9 columns]

## Libros mas vendidos

In [8]:
books = pd.read_csv('/kaggle/input/best-selling-books/best-selling-books.csv')
books

Book  \
0                                 A Tale of Two Cities   
1                  The Little Prince (Le Petit Prince)   
2             Harry Potter and the Philosopher's Stone   
3                             And Then There Were None   
4                       Dream of the Red Chamber (紅樓夢)   
..                                                 ...   
169                                           The Goal   
170                                     Fahrenheit 451   
171                                     Angela's Ashes   
172  The Story of My Experiments with Truth (સત્યના...   
173                              Bridget Jones's Diary   

                      Author(s) Original language  First published  \
0               Charles Dickens           English             1859   
1      Antoine de Saint-Exupéry            French             1943   
2                 J. K. Rowling           English             1997   
3               Agatha Christie           English             1939   
4                    Cao Xueqin           Chinese             1791   
..                          ...               ...              ...   
169         Eliyahu M. Goldratt           English             1984   
170                Ray Bradbury           English             1953   
171               Frank McCourt           English             1996   
172  Mohandas Karamchand Gandhi          Gujarati             1929   
173              Helen Fielding           English             1996   

     Approximate sales in millions               Genre  
0                            200.0  Historical fiction  
1                            200.0             Novella  
2                            120.0             Fantasy  
3                            100.0             Mystery  
4                            100.0         Family saga  
..                             ...                 ...  
169                           10.0                 NaN  
170                           10.0                 NaN  
171                           10.0                 NaN  
172                           10.0                 NaN  
173                           10.0                 NaN  

[174 rows x 6 columns]

In [9]:
books.describe()

First published  Approximate sales in millions
count       174.000000                     174.000000
mean       1962.522989                      30.097126
std          64.268737                      27.957985
min        1304.000000                      10.000000
25%        1947.000000                      14.000000
50%        1974.000000                      20.000000
75%        1995.000000                      36.300000
max        2018.000000                     200.000000

In [10]:
print(len(books['Genre'].unique()))
print(len(books['Author(s)'].unique()))
print(len(books['First published'].unique()))
print(len(books['Approximate sales in millions'].unique()))
print(len(books['Original language'].unique()))

81
157
95
47
16


In [11]:
books = books.drop(['First published'], axis=1) # too much indexes
# books = books.drop(['Approximate sales in millions'], axis=1) # too much indexes # objetive
books = books.drop(['Author(s)'], axis=1) # too much indexes
books = books.drop(['Book'], axis=1) # already id
books = books.drop(['Genre'], axis=1) # too much indexes

In [12]:
x=books

In [13]:
%%time

book_encoding=pd.DataFrame()
label=LabelEncoder()
for c in  x.columns:
    if(x[c].dtype=='object'):
        book_encoding[c]=label.fit_transform(x[c])
    else:
        book_encoding[c]=x[c]
        
book_encoding

CPU times: user 2.19 ms, sys: 1.14 ms, total: 3.33 ms
Wall time: 3.55 ms


Original language  Approximate sales in millions
0                    3                          200.0
1                    4                          200.0
2                    3                          120.0
3                    3                          100.0
4                    0                          100.0
..                 ...                            ...
169                  3                           10.0
170                  3                           10.0
171                  3                           10.0
172                  6                           10.0
173                  3                           10.0

[174 rows x 2 columns]

## Titanic

In [14]:
titanic = pd.read_csv('/kaggle/input/test-file/tested.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [15]:
print(len(titanic['Pclass'].unique()))
print(len(titanic['Name'].unique()))
print(len(titanic['SibSp'].unique()))
print(len(titanic['Parch'].unique()))
print(len(titanic['Ticket'].unique()))
print(len(titanic['Fare'].unique()))
print(len(titanic['Cabin'].unique()))
print(len(titanic['Embarked'].unique()))
print(len(titanic['Age'].unique()))

3
418
7
8
363
170
77
3
80


In [16]:
titanic = titanic.drop(['Name'], axis=1) # too much indexes
titanic = titanic.drop(['Ticket'], axis=1) # too much indexes
titanic = titanic.drop(['Fare'], axis=1) # too much indexes
titanic = titanic.drop(['Cabin'], axis=1) # too much indexes
titanic = titanic.drop(['Age'], axis=1) # too much indexes
titanic = titanic.drop(['PassengerId'], axis=1) # already an id

In [17]:
titanic.head(5)

Survived  Pclass     Sex  SibSp  Parch Embarked
0         0       3    male      0      0        Q
1         1       3  female      1      0        S
2         0       2    male      0      0        Q
3         0       3    male      0      0        S
4         1       3  female      1      1        S

In [18]:
x = titanic

In [19]:
%%time

titanic_encoding=pd.DataFrame()
label=LabelEncoder()
for c in  x.columns:
    if(x[c].dtype=='object'):
        titanic_encoding[c]=label.fit_transform(x[c])
    else:
        titanic_encoding[c]=x[c]
        
titanic_encoding

CPU times: user 4.26 ms, sys: 46 µs, total: 4.3 ms
Wall time: 4.13 ms


Survived  Pclass  Sex  SibSp  Parch  Embarked
0           0       3    1      0      0         1
1           1       3    0      1      0         2
2           0       2    1      0      0         1
3           0       3    1      0      0         2
4           1       3    0      1      1         2
..        ...     ...  ...    ...    ...       ...
413         0       3    1      0      0         2
414         1       1    0      0      0         0
415         0       3    1      0      0         2
416         0       3    1      0      0         2
417         0       3    1      1      1         0

[418 rows x 6 columns]

# One Hot Encoding

## Ingresos de personas adultas

In [20]:
adult_census_df = pd.read_csv('/kaggle/input/adult-census-income-dataset/adult.csv')
x = adult_census_df
adult_census_df.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

In [21]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

age: 73
workclass: 9
fnlwgt: 21648
education: 16
education.num: 16
marital.status: 7
occupation: 15
relationship: 6
race: 5
sex: 2
capital.gain: 119
capital.loss: 92
hours.per.week: 94
native.country: 42
income: 2


In [22]:
x = x.drop(['age'], axis=1) # too much indexes
x = x.drop(['fnlwgt'], axis=1) # too much indexes
x = x.drop(['education'], axis=1) # too much indexes
x = x.drop(['education.num'], axis=1) # another column fulfills this column
x = x.drop(['occupation'], axis=1) # too much indexes
x = x.drop(['capital.gain'], axis=1) # too much indexes
x = x.drop(['capital.loss'], axis=1) # too much indexes
x = x.drop(['hours.per.week'], axis=1) # too much indexes
x = x.drop(['native.country'], axis=1) # too much indexes

In [23]:
x.columns

Index(['workclass', 'marital.status', 'relationship', 'race', 'sex', 'income'], dtype='object')

In [24]:
%%time 

res = pd.DataFrame()
one=OneHotEncoder()

#Definimos las columnas que queremos convertir
columnasOneHotEncoder=one.fit_transform(x[x.columns])

print('res data set has got {} rows and {} columns'.format(res.shape[0],res.shape[1]))

#Lo convertimos en array
columnasOneHotEncoder=columnasOneHotEncoder.toarray()

#Creamos el dataFrame indicando las columnas con el método get_feature_names()
columnasOneHotEncoder=pd.DataFrame(columnasOneHotEncoder,columns=one.get_feature_names_out())

#Lo añadimos a nuestro dataFrame para que aparezca las nuevas columnas
res=pd.concat([x,columnasOneHotEncoder],axis=1)
res.head(3)

res data set has got 0 rows and 0 columns
CPU times: user 63.2 ms, sys: 14 ms, total: 77.3 ms
Wall time: 77.5 ms


workclass marital.status   relationship   race     sex income  workclass_?  \
0         ?        Widowed  Not-in-family  White  Female  <=50K          1.0   
1   Private        Widowed  Not-in-family  White  Female  <=50K          0.0   
2         ?        Widowed      Unmarried  Black  Female  <=50K          1.0   

   workclass_Federal-gov  workclass_Local-gov  workclass_Never-worked  \
0                    0.0                  0.0                     0.0   
1                    0.0                  0.0                     0.0   
2                    0.0                  0.0                     0.0   

   workclass_Private  workclass_Self-emp-inc  workclass_Self-emp-not-inc  \
0                0.0                     0.0                         0.0   
1                1.0                     0.0                         0.0   
2                0.0                     0.0                         0.0   

   workclass_State-gov  workclass_Without-pay  marital.status_Divorced  \
0                  0.0                    0.0                      0.0   
1                  0.0                    0.0                      0.0   
2                  0.0                    0.0                      0.0   

   marital.status_Married-AF-spouse  marital.status_Married-civ-spouse  \
0                               0.0                                0.0   
1                               0.0                                0.0   
2                               0.0                                0.0   

   marital.status_Married-spouse-absent  marital.status_Never-married  \
0                                   0.0                           0.0   
1                                   0.0                           0.0   
2                                   0.0                           0.0   

   marital.status_Separated  marital.status_Widowed  relationship_Husband  \
0                       0.0                     1.0                   0.0   
1                       0.0                     1.0                   0.0   
2                       0.0                     1.0                   0.0   

   relationship_Not-in-family  relationship_Other-relative  \
0                         1.0                          0.0   
1                         1.0                          0.0   
2                         0.0                          0.0   

   relationship_Own-child  relationship_Unmarried  relationship_Wife  \
0                     0.0                     0.0                0.0   
1                     0.0                     0.0                0.0   
2                     0.0                     1.0                0.0   

   race_Amer-Indian-Eskimo  race_Asian-Pac-Islander  race_Black  race_Other  \
0                      0.0                      0.0         0.0         0.0   
1                      0.0                      0.0         0.0         0.0   
2                      0.0                      0.0         1.0         0.0   

   race_White  sex_Female  sex_Male  income_<=50K  income_>50K  
0         1.0         1.0       0.0           1.0          0.0  
1         1.0         1.0       0.0           1.0          0.0  
2         0.0         1.0       0.0           1.0          0.0

## Libros mas vendidos

In [25]:
books = pd.read_csv('/kaggle/input/best-selling-books/best-selling-books.csv')
x = books
x.head()

Book                 Author(s)  \
0                      A Tale of Two Cities           Charles Dickens   
1       The Little Prince (Le Petit Prince)  Antoine de Saint-Exupéry   
2  Harry Potter and the Philosopher's Stone             J. K. Rowling   
3                  And Then There Were None           Agatha Christie   
4            Dream of the Red Chamber (紅樓夢)                Cao Xueqin   

  Original language  First published  Approximate sales in millions  \
0           English             1859                          200.0   
1            French             1943                          200.0   
2           English             1997                          120.0   
3           English             1939                          100.0   
4           Chinese             1791                          100.0   

                Genre  
0  Historical fiction  
1             Novella  
2             Fantasy  
3             Mystery  
4         Family saga

In [26]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

Book: 174
Author(s): 157
Original language: 16
First published: 95
Approximate sales in millions: 47
Genre: 80


In [27]:
x = x.drop(['Book'], axis=1) # already id
# books = books.drop(['Approximate sales in millions'], axis=1) # too much indexes # objetive
x = x.drop(['Author(s)'], axis=1) # too much indexes
x = x.drop(['First published'], axis=1) # too much indexes
x = x.drop(['Genre'], axis=1) # too much indexes

In [28]:
x.columns

Index(['Original language', 'Approximate sales in millions'], dtype='object')

In [29]:
%%time 

res = pd.DataFrame()
one=OneHotEncoder()

#Definimos las columnas que queremos convertir
columnasOneHotEncoder=one.fit_transform(x[x.columns])

print('res data set has got {} rows and {} columns'.format(res.shape[0],res.shape[1]))

#Lo convertimos en array
columnasOneHotEncoder=columnasOneHotEncoder.toarray()

#Creamos el dataFrame indicando las columnas con el método get_feature_names()
columnasOneHotEncoder=pd.DataFrame(columnasOneHotEncoder,columns=one.get_feature_names_out())

#Lo añadimos a nuestro dataFrame para que aparezca las nuevas columnas
res=pd.concat([x,columnasOneHotEncoder],axis=1)
res.head(3)

res data set has got 0 rows and 0 columns
CPU times: user 6.12 ms, sys: 12 µs, total: 6.13 ms
Wall time: 5.55 ms


Original language  Approximate sales in millions  Original language_Chinese  \
0           English                          200.0                        0.0   
1            French                          200.0                        0.0   
2           English                          120.0                        0.0   

   Original language_Czech  Original language_Dutch  \
0                      0.0                      0.0   
1                      0.0                      0.0   
2                      0.0                      0.0   

   Original language_English  Original language_French  \
0                        1.0                       0.0   
1                        0.0                       1.0   
2                        1.0                       0.0   

   Original language_German  Original language_Gujarati  \
0                       0.0                         0.0   
1                       0.0                         0.0   
2                       0.0                         0.0   

   Original language_Hindi  Original language_Italian  \
0                      0.0                        0.0   
1                      0.0                        0.0   
2                      0.0                        0.0   

   Original language_Japanese  Original language_Norwegian  \
0                         0.0                          0.0   
1                         0.0                          0.0   
2                         0.0                          0.0   

   Original language_Portuguese  Original language_Russian  \
0                           0.0                        0.0   
1                           0.0                        0.0   
2                           0.0                        0.0   

   Original language_Spanish  Original language_Swedish  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   

   Original language_Yiddish  Approximate sales in millions_10.0  \
0                        0.0                                 0.0   
1                        0.0                                 0.0   
2                        0.0                                 0.0   

   Approximate sales in millions_10.4  Approximate sales in millions_10.5  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   

   Approximate sales in millions_11.0  Approximate sales in millions_11.5  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   

   Approximate sales in millions_12.0  Approximate sales in millions_12.1  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   

   Approximate sales in millions_13.0  Approximate sales in millions_14.0  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   

   Approximate sales in millions_15.0  Approximate sales in millions_15.2  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   

   Approximate sales in millions_15.3  Approximate sales in millions_16.0  \
0                                 0.0                                 0.0   
1                                 0.0             

## Titanic

In [30]:
titanic = pd.read_csv('/kaggle/input/test-file/tested.csv')
x = titanic
x.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [31]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

PassengerId: 418
Survived: 2
Pclass: 3
Name: 418
Sex: 2
Age: 79
SibSp: 7
Parch: 8
Ticket: 363
Fare: 169
Cabin: 76
Embarked: 3


In [32]:
x = x.drop(['PassengerId'], axis=1) # already an id
x = x.drop(['Name'], axis=1) # too much indexes
x = x.drop(['Age'], axis=1) # too much indexes
x = x.drop(['Ticket'], axis=1) # too much indexes
x = x.drop(['Fare'], axis=1) # too much indexes
x = x.drop(['Cabin'], axis=1) # too much indexes

In [33]:
x.columns

Index(['Survived', 'Pclass', 'Sex', 'SibSp', 'Parch', 'Embarked'], dtype='object')

In [34]:
%%time 

res = pd.DataFrame()
one=OneHotEncoder()

#Definimos las columnas que queremos convertir
columnasOneHotEncoder=one.fit_transform(x[x.columns])

print('res data set has got {} rows and {} columns'.format(res.shape[0],res.shape[1]))

#Lo convertimos en array
columnasOneHotEncoder=columnasOneHotEncoder.toarray()

#Creamos el dataFrame indicando las columnas con el método get_feature_names()
columnasOneHotEncoder=pd.DataFrame(columnasOneHotEncoder,columns=one.get_feature_names_out())

#Lo añadimos a nuestro dataFrame para que aparezca las nuevas columnas
res=pd.concat([x,columnasOneHotEncoder],axis=1)
res.head(10)

res data set has got 0 rows and 0 columns
CPU times: user 7.52 ms, sys: 1.77 ms, total: 9.3 ms
Wall time: 8.71 ms


Survived  Pclass     Sex  SibSp  Parch Embarked  Survived_0  Survived_1  \
0         0       3    male      0      0        Q         1.0         0.0   
1         1       3  female      1      0        S         0.0         1.0   
2         0       2    male      0      0        Q         1.0         0.0   
3         0       3    male      0      0        S         1.0         0.0   
4         1       3  female      1      1        S         0.0         1.0   
5         0       3    male      0      0        S         1.0         0.0   
6         1       3  female      0      0        Q         0.0         1.0   
7         0       2    male      1      1        S         1.0         0.0   
8         1       3  female      0      0        C         0.0         1.0   
9         0       3    male      2      0        S         1.0         0.0   

   Pclass_1  Pclass_2  Pclass_3  Sex_female  Sex_male  SibSp_0  SibSp_1  \
0       0.0       0.0       1.0         0.0       1.0      1.0      0.0   
1       0.0       0.0       1.0         1.0       0.0      0.0      1.0   
2       0.0       1.0       0.0         0.0       1.0      1.0      0.0   
3       0.0       0.0       1.0         0.0       1.0      1.0      0.0   
4       0.0       0.0       1.0         1.0       0.0      0.0      1.0   
5       0.0       0.0       1.0         0.0       1.0      1.0      0.0   
6       0.0       0.0       1.0         1.0       0.0      1.0      0.0   
7       0.0       1.0       0.0         0.0       1.0      0.0      1.0   
8       0.0       0.0       1.0         1.0       0.0      1.0      0.0   
9       0.0       0.0       1.0         0.0       1.0      0.0      0.0   

   SibSp_2  SibSp_3  SibSp_4  SibSp_5  SibSp_8  Parch_0  Parch_1  Parch_2  \
0      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
1      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
2      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
3      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
4      0.0      0.0      0.0      0.0      0.0      0.0      1.0      0.0   
5      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
6      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
7      0.0      0.0      0.0      0.0      0.0      0.0      1.0      0.0   
8      0.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   
9      1.0      0.0      0.0      0.0      0.0      1.0      0.0      0.0   

   Parch_3  Parch_4  Parch_5  Parch_6  Parch_9  Embarked_C  Embarked_Q  \
0      0.0      0.0      0.0      0.0      0.0         0.0         1.0   
1      0.0      0.0      0.0      0.0      0.0         0.0         0.0   
2      0.0      0.0      0.0      0.0      0.0         0.0         1.0   
3      0.0      0.0      0.0      0.0      0.0         0.0         0.0   
4      0.0      0.0      0.0      0.0      0.0         0.0         0.0   
5      0.0      0.0      0.0      0.0      0.0         0.0         0.0   
6      0.0      0.0      0.0      0.0      0.0         0.0         1.0   
7      0.0      0.0      0.0      0.0      0.0         0.0         0.0   
8      0.0      0.0      0.0      0.0      0.0         1.0         0.0   
9      0.0      0.0      0.0      0.0      0.0         0.0         0.0   

   Embarked_S  
0         0.0  
1         1.0  
2         0.0  
3         1.0  
4         1.0  
5         1.0  
6         0.0  
7         1.0  
8         0.0  
9         1.0

# Hashing de Características

In [35]:
from sklearn.feature_extraction import FeatureHasher
np.set_printoptions(threshold=np.inf)

## Ingresos de personas Adultas

In [36]:
adult_census_df = pd.read_csv('/kaggle/input/adult-census-income-dataset/adult.csv')
x = adult_census_df
x.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

In [37]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

age: 73
workclass: 9
fnlwgt: 21648
education: 16
education.num: 16
marital.status: 7
occupation: 15
relationship: 6
race: 5
sex: 2
capital.gain: 119
capital.loss: 92
hours.per.week: 94
native.country: 42
income: 2


In [38]:
x = x.drop(['age'], axis=1) # extended index
x = x.drop(['fnlwgt'], axis=1) # extended index, not relevant
x = x.drop(['capital.gain'], axis=1) # extended index
x = x.drop(['capital.loss'], axis=1) # extended index
x = x.drop(['hours.per.week'], axis=1) # extended index
x = x.drop(['native.country'], axis=1) # extended index
x = x.drop(['education.num'], axis=1) # education.num already fulfills the function

In [39]:
x.columns

Index(['workclass', 'education', 'marital.status', 'occupation',
       'relationship', 'race', 'sex', 'income'],
      dtype='object')

In [40]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

workclass: 9
education: 16
marital.status: 7
occupation: 15
relationship: 6
race: 5
sex: 2
income: 2


In [41]:
# Convertir filas a diccionarios
data_dicts = x.to_dict(orient="records")

hashing=FeatureHasher(n_features= 20,input_type='string')
hash_train = hashing.transform(data_dicts)
print(hash_train.toarray()[:20, :20])

[[ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.  0.  0.
   1.  1.]
 [ 1.  1.  0. -1.  0.  0.  0.  0. -1.  0.  0.  1.  0.  0.  0. -1.

# Target Encoding

#### ¿Qué es?  
En esta técnica, cada categoría de una variable se reemplaza por una métrica derivada del objetivo (por ejemplo, la media del objetivo para cada categoría).
Es especialmente útil en problemas de clasificación o regresión.
#### Ventajas:
- Reduce la dimensionalidad: A diferencia de One-Hot Encoding, no genera nuevas columnas.
- Captura relaciones directas entre las categorías y el objetivo.
- Funciona bien con datos categóricos de alta cardinalidad.
#### Desventajas:
- Puede causar fuga de datos si no se hace correctamente (el objetivo futuro se filtra al modelo durante el entrenamiento).
- Es sensible a categorías con pocos datos (puede introducir ruido).

## Ingreso de personas Adultas

In [42]:
adult_census_df = pd.read_csv('/kaggle/input/adult-census-income-dataset/adult.csv')
x = adult_census_df
x['income'] = x['income'].map({'<=50K': 0, '>50K': 1})
x.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country  income  
0          4356              40  United-States       0  
1          4356              18  United-States       0  
2          4356              40  United-States       0  
3          3900              40  United-States       0  
4          3900              40  United-States       0

In [43]:
categorical_columns = ['workclass', 'education', 'marital.status', 'occupation', 'relationship', 'race', 'sex', 'native.country']
y = x[categorical_columns]
y.head()

workclass     education marital.status         occupation   relationship  \
0         ?       HS-grad        Widowed                  ?  Not-in-family   
1   Private       HS-grad        Widowed    Exec-managerial  Not-in-family   
2         ?  Some-college        Widowed                  ?      Unmarried   
3   Private       7th-8th       Divorced  Machine-op-inspct      Unmarried   
4   Private  Some-college      Separated     Prof-specialty      Own-child   

    race     sex native.country  
0  White  Female  United-States  
1  White  Female  United-States  
2  Black  Female  United-States  
3  White  Female  United-States  
4  White  Female  United-States

In [44]:
# Aplicar Target Encoding a cada columna categórica
for col in categorical_columns:
    mean_income_per_category = x.groupby(col)['income'].mean()
    x[col] = x[col].map(mean_income_per_category)

# Mostrar el dataset con Target Encoding aplicado
x = x[categorical_columns + ['income']]
x.head()

workclass  education  marital.status  occupation  relationship     race  \
0   0.104031   0.159509        0.085599    0.103635      0.103070  0.25586   
1   0.218673   0.159509        0.085599    0.484014      0.103070  0.25586   
2   0.104031   0.190235        0.085599    0.103635      0.063262  0.12388   
3   0.218673   0.061920        0.104209    0.124875      0.063262  0.25586   
4   0.218673   0.190235        0.064390    0.449034      0.013220  0.25586   

        sex  native.country  income  
0  0.109461        0.245835       0  
1  0.109461        0.245835       0  
2  0.109461        0.245835       0  
3  0.109461        0.245835       0  
4  0.109461        0.245835       0

#### Significado

Cada valor numérico en las nuevas columnas refleja la media del ingreso (en este caso, 0 o 1) para cada categoría. Por ejemplo, el valor 0.104031 en la columna workclass significa que la media de income para esa categoría específica en la columna workclass es aproximadamente 0.104031.

## Libros mas vendidos

In [45]:
books = pd.read_csv('/kaggle/input/best-selling-books/best-selling-books.csv')
x = books
x.head()

Book                 Author(s)  \
0                      A Tale of Two Cities           Charles Dickens   
1       The Little Prince (Le Petit Prince)  Antoine de Saint-Exupéry   
2  Harry Potter and the Philosopher's Stone             J. K. Rowling   
3                  And Then There Were None           Agatha Christie   
4            Dream of the Red Chamber (紅樓夢)                Cao Xueqin   

  Original language  First published  Approximate sales in millions  \
0           English             1859                          200.0   
1            French             1943                          200.0   
2           English             1997                          120.0   
3           English             1939                          100.0   
4           Chinese             1791                          100.0   

                Genre  
0  Historical fiction  
1             Novella  
2             Fantasy  
3             Mystery  
4         Family saga

In [46]:
categorical_columns = ['Author(s)', 'Original language', 'Genre']
y = x[categorical_columns]
y.head()

Author(s) Original language               Genre
0           Charles Dickens           English  Historical fiction
1  Antoine de Saint-Exupéry            French             Novella
2             J. K. Rowling           English             Fantasy
3           Agatha Christie           English             Mystery
4                Cao Xueqin           Chinese         Family saga

In [47]:
# Aplicar Target Encoding a cada columna categórica
for col in categorical_columns:
    mean_income_per_category = x.groupby(col)['Approximate sales in millions'].mean()
    x[col] = x[col].map(mean_income_per_category)

# Mostrar el dataset con Target Encoding aplicado
x = x[categorical_columns + ['Approximate sales in millions']]
x.head()

Author(s)  Original language   Genre  Approximate sales in millions
0  200.000000          30.274809   73.75                          200.0
1  200.000000          51.400000  200.00                          200.0
2   74.571429          30.274809   73.70                          120.0
3  100.000000          30.274809  100.00                          100.0
4  100.000000          35.000000  100.00                          100.0

## Titanic

In [48]:
titanic = pd.read_csv('/kaggle/input/test-file/tested.csv')
x = titanic
x.head()

PassengerId  Survived  Pclass  \
0          892         0       3   
1          893         1       3   
2          894         0       2   
3          895         0       3   
4          896         1       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                              Kelly, Mr. James    male  34.5      0      0   
1              Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                     Myles, Mr. Thomas Francis    male  62.0      0      0   
3                              Wirz, Mr. Albert    male  27.0      0      0   
4  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   

    Ticket     Fare Cabin Embarked  
0   330911   7.8292   NaN        Q  
1   363272   7.0000   NaN        S  
2   240276   9.6875   NaN        Q  
3   315154   8.6625   NaN        S  
4  3101298  12.2875   NaN        S

In [49]:
categorical_columns = ['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked']
y = x[categorical_columns]
y.head()

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0       3    male  34.5      0      0   7.8292        Q
1       3  female  47.0      1      0   7.0000        S
2       2    male  62.0      0      0   9.6875        Q
3       3    male  27.0      0      0   8.6625        S
4       3  female  22.0      1      1  12.2875        S

In [50]:
# Aplicar Target Encoding a cada columna categórica
for col in categorical_columns:
    mean_income_per_category = x.groupby(col)['Survived'].mean()
    x[col] = x[col].map(mean_income_per_category)

# Mostrar el dataset con Target Encoding aplicado
x = x[categorical_columns + ['Survived']]
x.head(10)

Pclass  Sex       Age     SibSp     Parch      Fare  Embarked  Survived
0  0.330275  0.0  0.000000  0.310954  0.305556  0.000000  0.521739         0
1  0.330275  1.0  0.400000  0.490909  0.305556  0.500000  0.325926         1
2  0.322581  0.0  0.000000  0.310954  0.305556  0.000000  0.521739         0
3  0.330275  0.0  0.333333  0.310954  0.305556  0.375000  0.325926         0
4  0.330275  1.0  0.625000  0.490909  0.538462  1.000000  0.325926         1
5  0.330275  0.0  0.000000  0.310954  0.305556  0.000000  0.325926         0
6  0.330275  1.0  0.400000  0.310954  0.305556  1.000000  0.521739         1
7  0.322581  0.0  0.333333  0.490909  0.538462  0.000000  0.325926         0
8  0.330275  1.0  0.538462  0.310954  0.305556  0.222222  0.392157         1
9  0.330275  0.0  0.176471  0.428571  0.305556  0.000000  0.325926         0

# Dataset con tematica de proyecto

## Label Encondig

In [51]:
spotify = pd.read_csv('/kaggle/input/30000-spotify-songs/spotify_songs.csv')
x = spotify
#pd.set_option('display.max_columns', None)
x.head()

track_id                                         track_name  \
0  6f807x0ima9a1j3VPbc7VN  I Don't Care (with Justin Bieber) - Loud Luxur...   
1  0r7CVbZTWZgbTCYdfa2P31                    Memories - Dillon Francis Remix   
2  1z1Hg7Vb0AhHDiEmnDE79l                    All the Time - Don Diablo Remix   
3  75FpbthrwQmzHlBJLuGdC7                  Call You Mine - Keanu Silva Remix   
4  1e8PAfcKUYoKkxPhrHqw4x            Someone You Loved - Future Humans Remix   

       track_artist  track_popularity          track_album_id  \
0        Ed Sheeran                66  2oCs0DGTsRO98Gh5ZSl2Cx   
1          Maroon 5                67  63rPSO264uRjW1X5E6cWv6   
2      Zara Larsson                70  1HoSmj2eLcsrR0vE9gThr4   
3  The Chainsmokers                60  1nqYsOef1yKKuGOVchbsk6   
4     Lewis Capaldi                69  7m7vv9wlQ4i0LFuJiE2zsQ   

                                    track_album_name track_album_release_date  \
0  I Don't Care (with Justin Bieber) [Loud Luxury...               2019-06-14   
1                    Memories (Dillon Francis Remix)               2019-12-13   
2                    All the Time (Don Diablo Remix)               2019-07-05   
3                        Call You Mine - The Remixes               2019-07-19   
4            Someone You Loved (Future Humans Remix)               2019-03-05   

  playlist_name             playlist_id playlist_genre playlist_subgenre  \
0     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop   
1     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop   
2     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop   
3     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop   
4     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop   

   danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.748   0.916    6    -2.634     1       0.0583        0.1020   
1         0.726   0.815   11    -4.969     1       0.0373        0.0724   
2         0.675   0.931    1    -3.432     0       0.0742        0.0794   
3         0.718   0.930    7    -3.778     1       0.1020        0.0287   
4         0.650   0.833    1    -4.672     1       0.0359        0.0803   

   instrumentalness  liveness  valence    tempo  duration_ms  
0          0.000000    0.0653    0.518  122.036       194754  
1          0.004210    0.3570    0.693   99.972       162600  
2          0.000023    0.1100    0.613  124.008       176616  
3          0.000009    0.2040    0.277  121.956       169093  
4          0.000000    0.0833    0.725  123.976       189052

In [52]:
for i in x.columns:
    print(f"{i}: {x[i].nunique()}")

track_id: 28356
track_name: 23449
track_artist: 10692
track_popularity: 101
track_album_id: 22545
track_album_name: 19743
track_album_release_date: 4530
playlist_name: 449
playlist_id: 471
playlist_genre: 6
playlist_subgenre: 24
danceability: 822
energy: 952
key: 12
loudness: 10222
mode: 2
speechiness: 1270
acousticness: 3731
instrumentalness: 4729
liveness: 1624
valence: 1362
tempo: 17684
duration_ms: 19785


In [53]:
categorical_columns = ['track_popularity', 'playlist_genre', 'playlist_subgenre', 'key', 'mode', 'track_artist']
x = x[categorical_columns]
x.head()

track_popularity playlist_genre playlist_subgenre  key  mode  \
0                66            pop         dance pop    6     1   
1                67            pop         dance pop   11     1   
2                70            pop         dance pop    1     0   
3                60            pop         dance pop    7     1   
4                69            pop         dance pop    1     1   

       track_artist  
0        Ed Sheeran  
1          Maroon 5  
2      Zara Larsson  
3  The Chainsmokers  
4     Lewis Capaldi

In [54]:
%%time

spotify_encoding=pd.DataFrame()
label=LabelEncoder()
for c in  x.columns:
    if(x[c].dtype=='object'):
        spotify_encoding[c]=label.fit_transform(x[c])
    else:
        spotify_encoding[c]=x[c]
        
spotify_encoding.head()

CPU times: user 42 ms, sys: 1.75 ms, total: 43.8 ms
Wall time: 43 ms


track_popularity  playlist_genre  playlist_subgenre  key  mode  \
0                66               2                  3    6     1   
1                67               2                  3   11     1   
2                70               2                  3    1     0   
3                60               2                  3    7     1   
4                69               2                  3    1     1   

   track_artist  
0          2782  
1          6084  
2         10416  
3          9215  
4          5402

## One Hot Encoding

In [55]:
x = spotify
categorical_columns = ['track_popularity', 'playlist_genre', 'playlist_subgenre', 'key', 'mode']
x = x[categorical_columns]

In [56]:
%%time 

res = pd.DataFrame()
one=OneHotEncoder()

#Definimos las columnas que queremos convertir
columnasOneHotEncoder=one.fit_transform(x[x.columns])

print('res data set has got {} rows and {} columns'.format(res.shape[0],res.shape[1]))

#Lo convertimos en array
columnasOneHotEncoder=columnasOneHotEncoder.toarray()

#Creamos el dataFrame indicando las columnas con el método get_feature_names()
columnasOneHotEncoder=pd.DataFrame(columnasOneHotEncoder,columns=one.get_feature_names_out())

#Lo añadimos a nuestro dataFrame para que aparezca las nuevas columnas
res=pd.concat([x,columnasOneHotEncoder],axis=1)
res.head(3)

res data set has got 0 rows and 0 columns
CPU times: user 85.8 ms, sys: 69.3 ms, total: 155 ms
Wall time: 152 ms


track_popularity playlist_genre playlist_subgenre  key  mode  \
0                66            pop         dance pop    6     1   
1                67            pop         dance pop   11     1   
2                70            pop         dance pop    1     0   

   track_popularity_0  track_popularity_1  track_popularity_2  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   

   track_popularity_3  track_popularity_4  track_popularity_5  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   

   track_popularity_6  track_popularity_7  track_popularity_8  \
0                 0.0                 0.0                 0.0   
1                 0.0                 0.0                 0.0   
2                 0.0                 0.0                 0.0   

   track_popularity_9  track_popularity_10  track_popularity_11  \
0                 0.0                  0.0                  0.0   
1                 0.0                  0.0                  0.0   
2                 0.0                  0.0                  0.0   

   track_popularity_12  track_popularity_13  track_popularity_14  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_15  track_popularity_16  track_popularity_17  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_18  track_popularity_19  track_popularity_20  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_21  track_popularity_22  track_popularity_23  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_24  track_popularity_25  track_popularity_26  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_27  track_popularity_28  track_popularity_29  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_30  track_popularity_31  track_popularity_32  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_33  track_popularity_34  track_popularity_35  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_36  track_popularity_37  track_popularity_38  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.0                  0.0                  0.0   

   track_popularity_39  track_popularity_40  track_popularity_41  \
0                  0.0                  0.0                  0.0   
1                  0.0                  0.0                  0.0   
2                  0.

## Hashing de Caracteristicas

## Target Encoding

In [57]:
x = spotify

categorical_columns = ['playlist_genre', 'playlist_subgenre', 'key', 'mode', 'energy', 'danceability', 'playlist_name', 'track_artist']
y = x[categorical_columns]

In [58]:
# Aplicar Target Encoding a cada columna categórica
for col in categorical_columns:
    mean_income_per_category = x.groupby(col)['track_popularity'].mean()
    x[col] = x[col].map(mean_income_per_category)

# Mostrar el dataset con Target Encoding aplicado
x = x[categorical_columns + ['track_popularity']]
x.head()

playlist_genre  playlist_subgenre        key       mode     energy  \
0        47.74487          52.079353  42.581648  42.709917  37.906250   
1        47.74487          52.079353  42.409212  42.709917  46.573529   
2        47.74487          52.079353  42.994015  42.173785  30.941176   
3        47.74487          52.079353  40.904535  42.709917  37.851064   
4        47.74487          52.079353  42.994015  42.709917  39.877551   

   danceability  playlist_name  track_artist  track_popularity  
0     45.795918      59.628571     74.231884                66  
1     43.760000      59.628571     53.042254                67  
2     36.931373      59.628571     57.178571                70  
3     39.655556      59.628571     57.699187                60  
4     34.333333      59.628571     83.714286                69